app/knight_training/train_all_knight.py

```diff
@@ -0,0 +1,27 @@
+def apply_armour(knight):
+    knight["protection"] = 0
+    for a in knight["armour"]:
+        knight["protection"] += a["protection"]
+
+
+def apply_weapon(knight):
+    knight["power"] += knight["weapon"]["power"]
+
+
+def apply_potion_if_exist(knight):
+    if knight["potion"] is not None:
+        if "power" in knight["potion"]["effect"]:
+            knight["power"] += knight["potion"]["effect"]["power"]
+
+        if "protection" in knight["potion"]["effect"]:
+            knight["protection"] += knight["potion"]["effect"]["protection"]
+
+        if "hp" in knight["potion"]["effect"]:
+            knight["hp"] += knight["potion"]["effect"]["hp"]
```

Comment:
These code duplications can be fixed, using `stats = ("protection", "power", "hp")` and looping on `stats`